### PR TITLE
add stuff needed to support ghc

### PR DESCRIPTION
**Pull request by ArsenArsen**

### CODE REVIEW

File: options/rtdl/generic/linker.hpp [hunk 1]
```diff
@@ -77,10 +77,10 @@ enum class HashStyle {
 using InitFuncPtr = void (*)();
 
 struct SharedObject {
-	SharedObject(const char *name, bool is_main_object,
+	SharedObject(frg::string<MemoryAllocator> name, bool is_main_object,
```
Comment (Geertiebear) [inline options/rtdl/generic/linker.hpp:80]: Maybe you could have a second constructor taking a ``const char *`` to clear up the ``frg::string`` instantiations you currently do across the code

File: options/posix/include/grp.h [hunk 2]
```diff
@@ -10,6 +10,7 @@ extern "C" {
 
 struct group {
 	char *gr_name;
+	char *gr_passwd;
```
Comment (avdgrinten) [inline options/posix/include/grp.h:13]: This is an ABI break

File: options/posix/include/pwd.h [hunk 3]
```diff
@@ -12,6 +12,7 @@ extern "C" {
 
 struct passwd {
 	char *pw_name;
+	char *pw_passwd;
```
Comment (avdgrinten) [inline options/posix/include/pwd.h:15]: This is an ABI break.

File: options/rtdl/generic/linker.cpp [hunk 4]
```diff
@@ -108,14 +111,53 @@ SharedObject *ObjectRepository::injectStaticObject(frg::string_view name,
...
+	};
+	auto doComponent = [&] (frg::string_view component) {
+		if (!component.size()) {
+			return;
+		}
+		if (component == ".") {
+			return;
+		}
+		if (component == "..") {
+			stack.pop();
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:132]: This is broken for symlinks.

File: options/rtdl/generic/linker.cpp [hunk 5]
```diff
@@ -462,9 +560,11 @@ void ObjectRepository::_parseDynamic(SharedObject *object) {
 		}
 	}
 
-	if(runpath_offset)
+	if(runpath_offset) {
 		object->runPath = reinterpret_cast<const char *>(object->baseAddress
 				+ object->stringTableOffset + *runpath_offset);
+		// TODO process tokens and add to global library list
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:566]: ?

Comment (ArsenArsen) [reply options/rtdl/generic/linker.cpp:566]: outdated, misunderstanding in the semantics of rpath

File: options/rtdl/generic/linker.cpp (hunk 4 above)
Comment (ArsenArsen) [reply options/rtdl/generic/linker.cpp:132]: the actual breakage to fix is in managarm-system, so this code could be removed

File: options/posix/include/pwd.h (hunk 3 above)
Comment (ArsenArsen) [reply options/posix/include/pwd.h:15]: what is the plan for resolving that?

File: options/rtdl/generic/linker.cpp (hunk 4 above)
Comment (avdgrinten) [reply options/rtdl/generic/linker.cpp:132]: Agreed. Path normalization should not be done here.

File: options/posix/include/pwd.h (hunk 3 above)
Comment (avdgrinten) [reply options/posix/include/pwd.h:15]: First, we should make sure that the structs are 100% complete now (e.g., compared to Linux). Then we merge it then but also announce the break.

File: options/posix/include/sys/time.h [hunk 6]
```diff
@@ -24,6 +24,21 @@ int timerisset(struct timeval *tvp);
 //TODO: implement timercmp as a macro
 //int timercmp(struct timeval *a, struct timeval *b, CMP);
 
+struct itimerval {
```
Comment (Geertiebear) [inline options/posix/include/sys/time.h:27]: this struct + defines should be in abi-bits

File: options/posix/include/bits/posix/stat.h [hunk 7]
```diff
@@ -20,5 +20,8 @@
 #define st_mtime st_mtim.tv_sec
 #define st_ctime st_ctim.tv_sec
 
+#define UTIME_NOW  ((1l << 30) - 1l)
```
Comment (Geertiebear) [inline options/posix/include/bits/posix/stat.h:23]: this also seems like something that should be in abi-bits

File: options/rtdl/generic/linker.cpp [hunk 8]
```diff
@@ -76,11 +78,13 @@ SharedObject *ObjectRepository::injectObjectFromDts(frg::string_view name,
...
 SharedObject *ObjectRepository::injectObjectFromPhdrs(frg::string_view name,
-		void *phdr_pointer, size_t phdr_entry_size, size_t num_phdrs, void *entry_pointer,
+		frg::string<MemoryAllocator> path, void *phdr_pointer,
+		size_t phdr_entry_size, size_t num_phdrs, void *entry_pointer,
 		uint64_t rts) {
 	__ensure(!_nameMap.get(name));
 
-	auto object = frg::construct<SharedObject>(getAllocator(), name.data(), true, rts);
+	auto object = frg::construct<SharedObject>(getAllocator(),
+		name.data(), path, true, rts);
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:87]: `std::move(path)`?

File: options/rtdl/generic/linker.cpp [hunk 9]
```diff
@@ -91,10 +95,12 @@ SharedObject *ObjectRepository::injectObjectFromPhdrs(frg::string_view name,
...
 
 SharedObject *ObjectRepository::injectStaticObject(frg::string_view name,
-		void *phdr_pointer, size_t phdr_entry_size, size_t num_phdrs, void *entry_pointer,
+		frg::string<MemoryAllocator> path, void *phdr_pointer,
+		size_t phdr_entry_size, size_t num_phdrs, void *entry_pointer,
 		uint64_t rts) {
 	__ensure(!_nameMap.get(name));
-	auto object = frg::construct<SharedObject>(getAllocator(), name.data(), true, rts);
+	auto object = frg::construct<SharedObject>(getAllocator(),
+		name.data(), path, true, rts);
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:103]: `std::move(path)`?

File: options/rtdl/generic/linker.cpp [hunk 10]
```diff
@@ -433,6 +492,7 @@ void ObjectRepository::_parseDynamic(SharedObject *object) {
 						<< ") is not implemented correctly!\e[39m"
 						<< frg::endlog;
 			break;
+		case DT_RPATH: // TODO: this is probably technically wrong
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:495]: Warn here that the precedence of RPATH is wrong.

File: options/rtdl/generic/linker.cpp [hunk 11]
```diff
@@ -158,11 +214,14 @@ SharedObject *ObjectRepository::requestObjectWithName(frg::string_view name,
 
 SharedObject *ObjectRepository::requestObjectAtPath(frg::string_view path, uint64_t rts) {
 	// TODO: Support SONAME correctly.
-	auto it = _nameMap.get(path);
+	auto lastSlash = path.find_last('/') + 1;
+	auto name = path.sub_string(lastSlash, path.size() - lastSlash);
+	auto it = _nameMap.get(name);
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:219]: What if there is no slash in `path`?

File: options/rtdl/generic/linker.cpp [hunk 12]
```diff
@@ -158,11 +214,14 @@ SharedObject *ObjectRepository::requestObjectWithName(frg::string_view name,
...
 	if(it)
 		return *it;
 
-	auto object = frg::construct<SharedObject>(getAllocator(), path.data(), false, rts);
+	auto object = frg::construct<SharedObject>(getAllocator(),
+		name.data(), path.data(), false, rts);
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:224]: Path is used again below but maybe if we use `object->path` instead, we can move here?

File: options/rtdl/generic/linker.cpp [hunk 13]
```diff
@@ -130,20 +134,72 @@ SharedObject *ObjectRepository::requestObjectWithName(frg::string_view name,
...
+	// TODO(arsen): this process can probably undergo heavy optimization, by
+	// preprocessing the rpath only once on parse
+	auto processRpath = [&] (frg::string_view &path) {
+		frg::string<MemoryAllocator> sPath { getAllocator() };
+		if (path.sub_string(0, 7) == "$ORIGIN") {
+			frg::string_view dirname = origin->path;
+			while (dirname.size() &&
+					dirname[dirname.size() - 1] != '/') {
+				dirname = { dirname.data(), dirname.size() - 1 };
+			}
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:146]: `while`?

File: options/rtdl/generic/linker.cpp [hunk 14]
```diff
@@ -130,20 +134,72 @@ SharedObject *ObjectRepository::requestObjectWithName(frg::string_view name,
 		return fd;
 	};
 
+	// TODO(arsen): this process can probably undergo heavy optimization, by
+	// preprocessing the rpath only once on parse
+	auto processRpath = [&] (frg::string_view &path) {
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:139]: Do not take a view by reference. (This is not changed inside the lambda, either.)

File: options/rtdl/generic/linker.cpp [hunk 15]
```diff
@@ -130,20 +134,72 @@ SharedObject *ObjectRepository::requestObjectWithName(frg::string_view name,
...
+		auto next = [&] () {
+			idx = rpath.find_first(':', start);
+		};
+		for (next(); idx < rpath.size(); next()) {
+			auto path = rpath.sub_string(start, idx - start);
+			start = idx + 1;
+			auto [fd_, fullPath] = processRpath(path);
+			if (fd_ != -1) {
+				fd = fd_;
+				fullName = fullPath;
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:178]: `std::move`.

File: options/rtdl/generic/linker.cpp [hunk 16]
```diff
@@ -130,20 +134,72 @@ SharedObject *ObjectRepository::requestObjectWithName(frg::string_view name,
...
+			}
+		}
+		if (fd == -1) {
+			auto path = rpath.sub_string(start, rpath.size() - start);
+			auto [fd_, fullPath] = processRpath(path);
+			if (fd_ != -1) {
+				fd = fd_;
+				fullName = fullPath;
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:186]: `std::move`.

File: options/rtdl/generic/linker.cpp [hunk 17]
```diff
@@ -130,20 +134,72 @@ SharedObject *ObjectRepository::requestObjectWithName(frg::string_view name,
...
+		}
+		if (sPath[sPath.size() - 1] != '/') {
+			sPath += '/';
+		}
+		sPath += name;
+		int fd = tryToOpen(sPath.data());
+		return frg::tuple { fd, std::move(sPath) };
+	};
+
+	frg::string<MemoryAllocator> fullName { getAllocator() };
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:163]: `chosenPath` /  `foundPath`?

File: options/rtdl/generic/linker.cpp (hunk 10 above)
Comment (avdgrinten) [reply options/rtdl/generic/linker.cpp]: By warn here, I meant: warn by emitting a warning over `mlibc::infoLogger`.

File: options/rtdl/generic/linker.cpp (hunk 13 above)
Comment (avdgrinten) [reply options/rtdl/generic/linker.cpp:148]: This is not fixed yet.

File: options/rtdl/generic/linker.cpp [hunk 18]
```diff
@@ -433,7 +497,9 @@ void ObjectRepository::_parseDynamic(SharedObject *object) {
 						<< ") is not implemented correctly!\e[39m"
 						<< frg::endlog;
 			break;
+		case DT_RPATH:
 		case DT_RUNPATH:
+			mlibc::infoLogger() << "\e[31mrtdl: RUNPATH not preferred over RPATH properly\e[39m" << frg::endlog;
```
Comment (avdgrinten) [inline options/rtdl/generic/linker.cpp:502]: This only needs to warn for `RPATH`. `RUNPATH` should work properly.

**Comment by avdgrinten** [comment]
LGTM now.

**Pull request merged**